### DISCUSSION TITLE
Migration to Godot 4.x

### DISCUSSION THREAD

**Discussion by charliewhitfield** [ivoyager/forum#6] [General]
It's happening!

I've created _godot4_ and _godot3_ branches in `ivoyager` and `planetarium` repositories, and started work on migration to Godot 4.1.1. For now, _master_ branch is frozen on our final Godot 3.x release, v0.0.15. Once I have something at least somewhat functional, the _master_ branch will start tracking our Godot 4.x work. My plan of attack is as follows:

1. Disable **everything** in project_builder.gd and planetarium.gd.
2. Do migration.
3. Fix compile errors in each file.
4. Re-enable, test and debug each individual system, one-at-a-time.

I'm pretty sure that almost every system will be broken, starting with our *.tsv data table import and asset file search & load systems. It may be a while before we are building the solar system.

I'm aware that there are HTML5 export problems still in Godot 4.1.1. I'll maintain the _godot3_ branches as long as needed for the web-based Planetarium, but I don't plan any 3.x support after that.

**Comment by 0xIonRod** [comment]
YES!
Looking forward the new version, I've failed migrating. At this point in time it makes sense to target to Godot 4.2, there are already betas, or at least keep in mind that.

I've spend a while looking into code, and here are my thoughts:

1. There is a nice plugin manager for Godot [gd-plug](https://github.com/imjp94/gd-plug), combined with git submodules, allows quite flawless decomposition of the project into several libraries
2. Here are my thoughts about the possible decomposition:

- File loading and IO operations like tsv/csv parsing
- GUI
- Visual rendering of solar system/planets
- Simulation core 
- Assets/Resource dowloader (new feature, highly needed, I'm planning to implement soon, happy to discuss)

Such a decomposition would be helpful for me, particularly File IO & Visual part.

Right now I'm working on a detailed 3D map of the Moon based on NASA CGI map, and I'm stuck with the task of detailed rendering of planets. As a simulation core I'm planning to make a GDextension for GMAT and rely on GMAT features, as that code has TRL-9.

To summarize, I'm Happy to collaborate on those tasks.

**Reply by charliewhitfield** [reply]
4.2 is at dev 1. But I'm sure it'll be beta before I get too far. I moved project discussion to a [new thread](https://github.com/orgs/ivoyager/discussions/7).

**Comment by charliewhitfield** [comment]
So, nearly every static function I coded in static/files.gd now has a Godot static equivalent in FileAccess or DirAccess or ResourceLoader. That's good. And the old File and Directory API is totally changed, so I have to jettison all my old code anyway...

**Comment by charliewhitfield** [comment]
After 24 commits in `ivoyager` and 16 commits in `planetarium` we have...&emsp;wait for it...&emsp;A boot screen!

**Comment by charliewhitfield** [comment]
We're building a solar system! I can see Earth and some orbit lines! For some reason Earth's radius is less than half of ISS's orbit radius (which isn't right). I can't select or move with the mouse. I haven't added any GUI yet.

**Comment by charliewhitfield** [comment]
Making progress! The [_godot4_ branch Planetarium](https://github.com/ivoyager/planetarium/tree/godot4) is running now with planets and moons. Body sizes, rotations and orbits look correct. And you can rotate the camera and move to targets with the mouse.

Still to do:
- ~Add back and repair GUI.~ Done!
- ~Migrate/repair the game save/load system (used by Project Template).~ Done!
- Figure out why we have no illumination from the Sun (only ambient light).
- Get asteroids working again with a new set of binaries (need to migrate [ivbinary_maker](https://github.com/ivoyager/ivbinary_maker) to do that).
- Repair the FragmentIdentifier system (used to identify shader-derived screen elements like orbit lines and asteroids).
- Fix Saturn Ring's color (it's very exaggerated).

**Comment by 0xIonRod** [comment]
Great work, FYI I'm following the progress, thank you so much for posting!

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------

**Comment by charliewhitfield** [comment]
I've re-generated our asteroid binary data using a Godot 4-migrated ivbinary_maker. There's good news and bad news...

- The bad news is that the shader-based asteroid points are broken.
- The good news is that the MultiMesh-based orbit lines are working. And working MUCH better than in Godot 3.x! Specifically, I can show orbit lines for ~75,000 asteroids (at 100 vertexes per orbit) on my laptop and it doesn't hang at all! This many would hang my laptop pretty seriously in Godot 3.x project builds. So there is certainly something working more efficiently in the new Godot 4 rendering.

The development replacement `asteroid_binaries` directory for Godot 4.x builds can be found [here](https://github.com/ivoyager/non_release_assets/releases/tag/2023-08-12).

**Comment by charliewhitfield** [comment]
I've merged _godot4_ branch into _master_ branch in all repositories. So this is now our official development build.

HOWEVER, we still have serious regressions. I'll update these as they are fixed (and other changes) in the changelog [here](https://github.com/ivoyager/ivoyager/blob/master/CHANGELOG.md).

Even after migration regressions are fixed, I'll have to maintain the _godot3_ branch until Godot 4.x can make a functional HTML5 export for our web-based planetarium. 

I think now I will take a break and play Baldur's Gate 3 for a while...

**Comment by 0xIonRod** [comment]
Hopefully HTML5 export will work better in the 4.2 version.

In the meantime I'm confirming that I was able to run current version on Linux and Godot 4.1.1

**Comment by charliewhitfield** [comment]
Making progress on the remaining migration regressions:

- The Sun is illuminating things again after turning light_energy up from 1.5 to 2.5e11.
- Asteroid points are visible again after recoding shaders to use new CUSTOM0, 1, 2 channels. (This fixes my greatest coding hack ever: using Mesh NORMAL, UV, UV2, etc., to pass asteroid orbital parameters.)

A few more listed in the [changelog](https://github.com/ivoyager/ivoyager/blob/master/CHANGELOG.md). Then I'll release v0.0.16.

**Comment by charliewhitfield** [comment]
All known migration regressions are fixed!

The new Saturn Rings shader is not just fixed, but massively improved. It can sample many LOD levels simultaneously when viewing up close at a steep angle. So we have super detail (the '1D' textures are 13177 texels) and zero visible aliasing! And I fixed an existing bug that prevented correct backscatter/forwardscatter color sampling. The new phase-angle effects are wonderful!

_master_ branch has the updates. Requires new assets [just uploaded](https://github.com/ivoyager/non_release_assets/releases).

The last bug I want to fix is pre-migration and it is driving me crazy. We have no shadows. For heaven's sake, what am I missing???!!!

**Comment by charliewhitfield** [comment]
Our first Godot 4 release v0.0.16 is out!

Expect v0.0.17 to follow very soon. This version will have major project restructuring (discussed [here](https://github.com/orgs/ivoyager/discussions/7)) with addons/ivoyager_core as an editor plugin.

**Comment by charliewhitfield** [comment]
Update:

The current _master_ branch (v0.0.18.dev) works great as a Windows export. 

The exported HTML5 is still non-functional as of Godot 4.2.beta2. Here's the Godot [issue](https://github.com/godotengine/godot/issues/83899).

My plan is to release v0.0.18 after Godot 4.2 official release, hopefully with a fix for HTML5. Then we'll will move into official "beta" stage with v0.1.

**Comment by charliewhitfield** [comment]
We have a working web app!

~Well, the lighting is buggy. But everything else seems to be working at first glance. If anyone wants to try it:~

[Edit:] New link to better debug build: https://www.ivoyager.dev/app-debug/planetarium-debug.html
